opal/rocm/common_rocm.c

```diff
@@ -0,0 +1,319 @@
+    if (OPAL_SUCCESS != ret) {
+        opal_output(0, "Failed to find hipStreamSynchronize\n");
+        dlclose(hip_handle);
+        return OPAL_ERROR;
+    }
+    *(void **) (&hip_funcs.hipStreamSynchronize) = ptr;
+
+    ret = opal_dl_lookup (hip_handle, "hipGetErrorString", &ptr, &str);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(0, "Failed to find hipGetErrorString\n");
+        dlclose(hip_handle);
+        return OPAL_ERROR;
+    }
+    *(void **) (&hip_funcs.hipGetErrorString) = ptr;
+
+    ret = opal_dl_lookup (hip_handle, "hipPointerGetAttributes", &ptr, &str);
+    if (OPAL_SUCCESS != ret) {
+        opal_output(0, "Failed to find hipPointerGetAttributes\n");
+        dlclose(hip_handle);
+        return OPAL_ERROR;
+    }
+    *(void **) (&hip_funcs.hipPointerGetAttributes) = ptr;
+
+    return OPAL_SUCCESS;
+}
+
+static void hip_dl_finalize(void)
+{
+    hipError_t err = hip_funcs.hipStreamDestroy(&common_rocm_MemcpyStream);
+    if (hipSuccess != err) {
+        opal_output(0, "hip_dl_finalize: error while destroying the hipStream\n");
+    }
+    dlclose(hip_handle);
+}
+
+
+static void mca_common_rocm_support_init(void)
+{
+    if (rocm_initialized) {
+        return;
+    }
+
+    OBJ_CONSTRUCT(&common_rocm_init_lock, opal_mutex_t);
+    OPAL_THREAD_LOCK(&common_rocm_init_lock);
+
+    if (rocm_initialized) {
+        return;
+    }
+
+    if (!OPAL_HAVE_DL_SUPPORT) {
+        opal_output(0, "Can not open libamdhip64.so without dlopen support\n");
+        OPAL_THREAD_UNLOCK(&common_rocm_init_lock);
+        return;
+    }
+    if (OPAL_SUCCESS != hip_dl_init()) {
+        opal_output(0, "Could not open libamdhip64.so. Please check your LD_LIBRARY_PATH\n");
+        OPAL_THREAD_UNLOCK(&common_rocm_init_lock);
+        return;
+    }
+    rocm_initialized = true;
+    opal_rocm_runtime_initialized = true;
+
+    mca_common_rocm_register_mca_variables();
+
+    hipError_t err = hip_funcs.hipStreamCreate(&common_rocm_MemcpyStream);
+    if (hipSuccess != err) {
+        opal_output(0, "Could not create hipStream, err=%d %s\n",
+                err, hip_funcs.hipGetErrorString(err));
+    }
+    OPAL_THREAD_UNLOCK(&common_rocm_init_lock);
+}
+
+void mca_common_rocm_convertor_init(opal_convertor_t *convertor, const void *pUserBuf)
+{
+    /* Only do the initialization on the first GPU access */
+    if (!rocm_initialized) {
+        mca_common_rocm_support_init();
+    }
+
+    convertor->cbmemcpy = (memcpy_fct_t) &mca_common_rocm_memcpy;
+    // set convertor->stream as well ? Not right now.
+
+    hipPointerAttribute_t attr;
+    hipError_t err =  hip_funcs.hipPointerGetAttributes(&attr, pUserBuf);
+    if (hipSuccess != err) {
+        return;
+    }
+    if (hipMemoryTypeDevice == attr.memoryType || hipMemoryTypeUnified == attr.memoryType) {
+        convertor->flags |= CONVERTOR_ROCM;
+    }
+}
+
+bool mca_common_rocm_check_bufs(char *dst, char *src)
```

Comment:
This is actually not done this way because of the cuda precedence set, but because it is the interface used at the moment in the datatype code:

https://github.com/open-mpi/ompi/blob/main/opal/datatype/opal_datatype_copy.c#L112

Reply:
fair enough; it seems like the DDT code could also change to just check both buffers, but I'm not going to fight here.